### PR TITLE
Settings rework

### PR DESCRIPTION
**Pull request by dydzio0614**
Replacement for #1432 

Moves settings to configurable interface, adds bunch of new settings and proper placeholder config values for unmerged stuff from other PR's. Remembers last used tab and opens it if possible. Comes with new lines to translate.

### CODE REVIEW

**Comment by heroesiiifan** [comment]
Oh god, the settings button is terrible. 😂
Do you need something there?

h3_button_vcmi_on
h3_button_vcmi_off

**Comment by dydzio0614** [comment]
So this is ready - video shows how it looks for now: https://streamable.com/v918cl

Three annoyances that are still in the code, not sure what to do with them:
1. Duplication of SetBoolSetting etc. static functions in files responsible for different tabs
2. Main window class has little code duplication from WindowBase
3. Dependency on battle interface

**Comment by IvanSavenko** [comment]
Based on video, will check code a bit later:
- I still think that we should remove "other" group since this is basically empty screen. Perhaps even just in json - disable button and remove all options from that page, so that tab can be restored later if we need more place for options.
- Text on buttons should say "General", "Adventure Map"/"Map", "Battle", without "Options" part. Too long & redundant.
- UI... still needs improvements... If these improvements are difficult for you - I am OK with merging as it for now, and then I'll look into improving it sometime before 1.2 release.

Also, which window is opened in combat? Your new UI? Will it directly open "Battle" tab?

**Comment by dydzio0614** [comment]
> * I still think that we should remove "other" group since this is basically empty screen. Perhaps even just in json - disable button and remove all options from that page, so that tab can be restored later if we need more place for options.

I consider leaving at is for following reasons:
1. Town things are sorta separate logically from core adventure map stuff
2. Labels for these particular options are kinda long, may be even longer in other languages - tab with more space may be better
3. Moving options forward and back between tabs makes mess in players memories and translatable string IDs if naming them after tabs. When new tab will be needed due to lack of space it will also probably contain small amount of options. And discouraging people from "flooding the tabs excessively just to avoid new tab creation" would be probably good idea.

What do you think?

> * Text on buttons should say "General", "Adventure Map"/"Map", "Battle", without "Options" part. Too long & redundant.

I kept up with original game's style for consistency and reused two original window titles, that may be minor improvement for people playing VCMI in language that we do not have in translations etc.

> * UI... still needs improvements... If these improvements are difficult for you - I am OK with merging as it for now, and then I'll look into improving it sometime before 1.2 release.

Yeah, i suck at this

File: client/battle/BattleInterfaceClasses.h [hunk 1]
```diff
@@ -14,6 +14,7 @@
 #include "../../lib/FunctionList.h"
 #include "../../lib/battle/BattleHex.h"
 #include "../windows/CWindowObject.h"
+#include "gui/InterfaceObjectConfigurable.h"
```
Comment (IvanSavenko) [inline client/battle/BattleInterfaceClasses.h:17]:
```suggestion

```

File: client/windows/settings/SettingsMainContainer.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	if(CSH->isGuest())
+	{
+		loadButton->block(true);
+		saveButton->block(true);
+		restartButton->block(true);
+	}
+
+	int defaultTabIndex = 0;
+	if(settings["general"]["lastSettingsTab"].isNumber())
+		defaultTabIndex = settings["general"]["lastSettingsTab"].Integer();
```
Comment (IvanSavenko) [inline client/windows/settings/SettingsMainContainer.cpp:68]: I think that if window is opened from battle we should unconditionally open battle tab instead

File: client/battle/BattleInterface.cpp [hunk 3]
```diff
@@ -97,7 +97,17 @@ BattleInterface::BattleInterface(const CCreatureSet *army1, const CCreatureSet *
 
 	CCS->musich->stopMusic();
 	setAnimationCondition(EAnimationEvents::OPENING, true);
-	battleIntroSoundChannel = CCS->soundh->playSoundFromSet(CCS->soundh->battleIntroSounds);
+
+	GH.pushInt(windowObject);
+	windowObject->blockUI(true);
+	windowObject->updateQueue();
+
+	if(settings["gameTweaks"]["skipBattleIntroMusic"].Bool())
```
Comment (IvanSavenko) [inline client/battle/BattleInterface.cpp:105]: Maybe move everything from this point and to the end of method into a separate method, e.g. playIntroSound?

File: client/windows/CWindowObject.cpp [hunk 4]
```diff
@@ -59,7 +59,7 @@ CWindowObject::CWindowObject(int options_, std::string imageName):
 	options(options_),
 	background(createBg(imageName, options_ & PLAYER_COLORED))
 {
-	assert(parent == nullptr); //Safe to remove, but windows should not have parent
+//	assert(parent == nullptr); //Safe to remove, but windows should not have parent
```
Comment (IvanSavenko) [inline client/windows/CWindowObject.cpp:62]: Why was this necessary?

File: client/battle/BattleInterface.cpp [hunk 5]
```diff
@@ -766,3 +741,11 @@ void BattleInterface::executeOnAnimationCondition( EAnimationEvents event, bool
...
 	awaitingEvents.push_back({action, event, state});
 }
+
+void BattleInterface::setBattleQueueVisibility(bool visible)
+{
+	if(visible)
+		windowObject->showQueue();
+	else
+		windowObject->hideQueue();
+}
```
Comment (IvanSavenko) [inline client/battle/BattleInterface.cpp:751]: Obviously minor, but please keep an empty line in the end of file
```suggestion
}

```

File: client/windows/settings/SettingsMainContainer.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+#pragma once
+
+#include "widgets/ObjectLists.h"
+#include "widgets/Buttons.h"
+#include "gui/InterfaceObjectConfigurable.h"
+#include "gui/CGuiHandler.h"
+#include "battle/BattleInterface.h"
+
+class SettingsMainContainer : public InterfaceObjectConfigurable
```
Comment (IvanSavenko) [inline client/windows/settings/SettingsMainContainer.h:19]: Maybe rename to SettingsWindow?

File: client/windows/settings/SettingsMainContainer.cpp [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+ * SettingsMainContainer.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#include <SDL_events.h>
```
Comment (IvanSavenko) [inline client/windows/settings/SettingsMainContainer.cpp:11]: No need for SDL from what I see.
Also, for includes please use following order:
1) StdInc.h
2) Main include (.h file with the same name as .cpp)
3) VCMI includes
4) Third-party includes
```suggestion
```

File: client/windows/settings/SettingsMainContainer.cpp [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+		saveButton->block(true);
+		restartButton->block(true);
+	}
+
+	int defaultTabIndex = 0;
+	if(settings["general"]["lastSettingsTab"].isNumber())
+		defaultTabIndex = settings["general"]["lastSettingsTab"].Integer();
+
+	parentBattleInterface = parentBattleUi;
+	tabContentArea = std::make_shared<CTabbedInt>(std::bind(&SettingsMainContainer::createTab, this, _1), Point(0, 40), defaultTabIndex);
```
Comment (IvanSavenko) [inline client/windows/settings/SettingsMainContainer.cpp:71]: Lambdas are preferred to std::bind

File: client/windows/settings/BattleOptionsTab.h [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+#pragma once
+
+#include "../../gui/InterfaceObjectConfigurable.h"
+#include "../../battle/BattleInterface.h"
```
Comment (IvanSavenko) [inline client/windows/settings/BattleOptionsTab.h:13]: Prefer forward-declarations to includes in headers
```suggestion

class BattleInterface;
```

File: client/windows/settings/SettingsMainContainer.h [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ */
+
+#pragma once
+
+#include "widgets/ObjectLists.h"
+#include "widgets/Buttons.h"
+#include "gui/InterfaceObjectConfigurable.h"
+#include "gui/CGuiHandler.h"
+#include "battle/BattleInterface.h"
```
Comment (IvanSavenko) [inline client/windows/settings/SettingsMainContainer.h:17]: Prefer forward declarations in headers
```suggestion
#include "gui/InterfaceObjectConfigurable.h"

class CTabbedInt;
class BattleInterface;
```

File: client/windows/settings/GeneralOptionsTab.cpp [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+#include "StdInc.h"
+
+#include "GeneralOptionsTab.h"
+
+#include <SDL_surface.h>
+#include <SDL_rect.h>
```
Comment (IvanSavenko) [inline client/windows/settings/GeneralOptionsTab.cpp:15]: Unused I think?
```suggestion
```

File: client/windows/settings/GeneralOptionsTab.cpp [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+	widget<CLabel>("resolutionLabel")->setText(resText);
+}
+
+void GeneralOptionsTab::setFullscreenMode(bool on)
+{
+	fillSelectableResolutions();
+
+	const auto & screenRes = settings["video"]["screenRes"];
+	const Point desiredResolution(screenRes["width"].Integer(), screenRes["height"].Integer());
+	const Point currentResolution(screen->w, screen->h);
```
Comment (IvanSavenko) [inline client/windows/settings/GeneralOptionsTab.cpp:155]:
```suggestion
	const Point currentResolution = GH.screenDimensions();
```

**Review by IvanSavenko — COMMENTED** [review]
I have (largely) ignored formatting & style, if your IDE (Clion was it?) can be integrated with clang-format & clang-tidy then you might want to run these tools on your code:
- one-liner lambdas
- C-style-cast
- std::bind usage

**Comment by IvanSavenko** [comment]
> 1. Town things are sorta separate logically from core adventure map stuff

I think should be OK to put in "General". I'll add grouping to background, and probably add subheaders, including "Town Window" header (similar to how H3 has "Auto-Combat Options" and "Creature Info" subheaders in battle window).

> 2. Labels for these particular options are kinda long, may be even longer in other languages - tab with more space may be better

Basically the same length as longest text in "Adventure Map" tab. Don't see a problem. And if it is - I would rather consider shorter name. It is too long right now anyway

> 3. Moving options forward and back between tabs makes mess in players memories and translatable string IDs if naming them after tabs. There is possibility that more fitting options will deserve space in adventure tab more. When new tab will be needed due to lack of space it will also probably contain small amount of options. And discouraging people from "flooding the tabs excessively just to avoid new tab creation" would be probably good idea.

Valid concern, but
- These are not options that are likely to be accessed often (like quick combat)
- These options should not be just dropped someplace else, but grouped with similar options in the same visual block (and later if redesign is needed - move entire block)
- I don't expect any major rearragement every release or something like that. This would be extremely rare event.
- Current screen is just too empty. And it can be said for all tabs. Compare its information density with H3 settings.

> I kept up with original game's style for consistency and reused two original window titles, that may be minor improvement for people playing VCMI in language that we do not have in translations etc.

H3 uses that text for window title, not tab names. As for translations - half of text will be missing anyway + with 1.2 we'll cover majority of our playerbase with translation support anyway

File: client/windows/settings/SettingsMainContainer.cpp (hunk 2 above)
Comment (dydzio0614) [reply client/windows/settings/SettingsMainContainer.cpp:72]: done

File: client/battle/BattleInterface.cpp (hunk 3 above)
Comment (dydzio0614) [reply client/battle/BattleInterface.cpp:110]: done

File: client/windows/CWindowObject.cpp (hunk 4 above)
Comment (dydzio0614) [reply client/windows/CWindowObject.cpp]: leftover from early progress

File: client/battle/BattleInterface.cpp (hunk 5 above)
Comment (dydzio0614) [reply client/battle/BattleInterface.cpp]: done

File: client/windows/settings/SettingsMainContainer.cpp (hunk 8 above)
Comment (dydzio0614) [reply client/windows/settings/SettingsMainContainer.cpp:75]: done

File: client/windows/settings/BattleOptionsTab.h (hunk 9 above)
Comment (dydzio0614) [reply client/windows/settings/BattleOptionsTab.h]: done

File: client/windows/settings/SettingsMainContainer.h (hunk 10 above)
Comment (dydzio0614) [reply client/windows/settings/SettingsMainContainer.h]: done

File: client/windows/settings/GeneralOptionsTab.cpp (hunk 11 above)
Comment (dydzio0614) [reply client/windows/settings/GeneralOptionsTab.cpp]: removed after using GH.screenDimensions()

File: client/windows/settings/GeneralOptionsTab.cpp (hunk 12 above)
Comment (dydzio0614) [reply client/windows/settings/GeneralOptionsTab.cpp]: done

File: client/battle/BattleInterfaceClasses.h (hunk 1 above)
Comment (dydzio0614) [reply client/battle/BattleInterfaceClasses.h]: done

File: client/windows/settings/SettingsMainContainer.h (hunk 6 above)
Comment (dydzio0614) [reply client/windows/settings/SettingsMainContainer.h]: Decided on SettingsMainWindow

**Comment by nullkiller** [comment]
Are these new png images free to use or kind of ripped assets from HoMM3?

**Comment by dydzio0614** [comment]
no, they are our existing assets, and golden border got made by Andruids

File: client/windows/settings/SettingsMainWindow.cpp [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+	}
+
+	int defaultTabIndex = 0;
+	if(parentBattleUi != nullptr)
+		defaultTabIndex = 2;
+	else if(settings["general"]["lastSettingsTab"].isNumber())
+		defaultTabIndex = settings["general"]["lastSettingsTab"].Integer();
+
+	parentBattleInterface = parentBattleUi;
+	tabContentArea = std::make_shared<CTabbedInt>(std::bind(&SettingsMainWindow::createTab, this, _1), Point(0, 40), defaultTabIndex);
```
Comment (IvanSavenko) [inline client/windows/settings/SettingsMainWindow.cpp:74]: Don't see any point in hardcoding tab position.
If goal is to ensure that tabs buttons won't overlap with tabs content then this can be achieved in json.
And without hardcoded position it might be possible to (for example) move tab buttons into another location purely via json edit.

File: client/windows/settings/SettingsMainWindow.cpp [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+#include "CPlayerInterface.h"
+#include "CServerHandler.h"
+
+
+SettingsMainWindow::SettingsMainWindow(BattleInterface * parentBattleUi) : InterfaceObjectConfigurable()
+{
+	OBJ_CONSTRUCTION_CAPTURING_ALL_NO_DISPOSE;
+
+	const JsonNode config(ResourceID("config/widgets/settings/settingsMainContainer.json"));
+	addCallback("activateSettingsTab", [this](int tabId) { openTab(tabId); });
```
Comment (IvanSavenko) [inline client/windows/settings/SettingsMainWindow.cpp:37]: Just thinking aloud, feel free to ignore:
if we were to register all callbacks for all tabs in one place (e.g. by creating SettingsMainWindow::registerCallbacks() method) we can allow moving UI elements between tabs without changes in code.

**Review by IvanSavenko — COMMENTED** [review]
Looks OK code-wise, other than that - see https://github.com/vcmi/vcmi/pull/1540#issuecomment-1433855552

**Comment by dydzio0614** [comment]
> Looks OK code-wise, other than that - see [#1540 (comment)](https://github.com/vcmi/vcmi/pull/1540#issuecomment-1433855552)

What is left unimplemented from that comment?

**Comment by IvanSavenko** [comment]
> What is left unimplemented from that comment?

"Other" tab still exists?

I still think that we should remove it, but this is not a blocker.

**Review by IvanSavenko — APPROVED** [review]

**Comment by dydzio0614** [comment]
@IvanSavenko other tab is commented in json

**Comment by IvanSavenko** [comment]
I think I'll merge PR's from @dydzio0614 tomorrow if this is fine with everyone.
Window needs better UX design, will look into it later myself, functionality-wise looks OK.

Will also merge other PR's since they were largely blocked by this PR:
- #1547 
- #1545 
- #1429
